### PR TITLE
Gedis with ZDB backend #24

### PR DESCRIPTION
**Pull request by BolaNasr**
# How to run Notary 
## in kosmos run the following: 
```python
j.builder.runtimes.lua.lua_rocks_install() 
j.builder.db.zdb.start() 
zdb_cl = j.clients.zdb.client_admin_get()  
zdb_cl = zdb_cl.namespace_new("notary_namespace", secret="1234")
bcdb = j.data.bcdb.new(zdbclient=zdb_cl, name="notary_bcdb")
bcdb.models_add("/sandbox/code/github/threefoldtech/digitalmeX/packages/notary/models")
server = j.servers.gedis.configure(host="0.0.0.0", port=8888)
server.actor_add('/sandbox/code/github/threefoldtech/digitalmeX/packages/notary/actors/notary_actor.py')
server.models_add(models=bcdb.models.values())
server.save()
server.start()                                          
```
## using Tmux
```
cd /sandbox/code/github/threefoldtech/digitalmeX/packages/notary

moonc . &&lapis server
```

# How to use it

### using gedis

```
gedis_client = j.clients.gedis.get("notary", port=8888) 
gedis_client.cmds.notary_actor.register("123", "aa==","{{CONTENT_SIGNATURE}}") 
```
### using curl 
```
curl -d '{"threebot_id":"123","content":"aa==", "content_signature":"{{CONTENT_SIGNATURE}}"}' -H "Content-Type: application/json" -X POST http://localhost:8080/register
```
### using gedis

```
gedis_client = j.clients.gedis.get("notary", port=8888) 
gedis_client.cmds.notary_actor.get("21491a8b3d97ad5b705e3be26e7") 
```

### from browser
```
http://localhost:8080/get?hash="21491a8b3d97ad5b705e3be26e7"
```

### CODE REVIEW

**Comment by BolaNasr** [comment]
in progress : zdb builder

**Comment by zaibon** [comment]
`zdb_cl = j.clients.zdb.start_test_instance()`
This cannot be the way to start a bcdb server, it clearly state this is a test instance.

File: packages/notary/actors/notary_actor.py [hunk 1]
```diff
@@ -11,10 +11,17 @@ class notary_actor(JSBASE):
     def start_page(self):
         return "welcome to https://github.com/threefoldtech/home/issues/135"
 
-    def register(self, robot_id, key, content, content_signature, schema_out):
+    def bcdb_get(self):
+        zdb_cl = j.clients.zdb.start_test_instance()
```
Comment (zaibon) [inline packages/notary/actors/notary_actor.py]: You should not start bcdb every time you do need to get an client instance to it.
BCDB should be started when gedis starts, and the you just need to have one client instance to be used here in the gedis handlers. 
Same remark for the line 18, you should not always add the models again.

File: packages/notary/actors/notary_actor.py [hunk 2]
```diff
@@ -24,8 +31,17 @@ def register(self, robot_id, key, content, content_signature, schema_out):
         ```
         """
         out = schema_out.new()
-        out.message = "your register info is robot_id = {} , key = {}, content = {} ,content_signature ={}".format(
-            robot_id, key, content, content_signature)
+        bcdb = self.bcdb_get()
```
Comment (zaibon) [inline packages/notary/actors/notary_actor.py]: Here we need to verify the signature before saving the content.
Then the return value should also be the blak2b hash of the content.

File: packages/notary/actors/notary_actor.py [hunk 3]
```diff
@@ -38,21 +54,29 @@ def get(self, key, schema_out):
...
+        for model in bcdb.get_all():
+            if model.key == key:
+                return model
+
+        out.message = "this key doesn't exist"
         return out
 
-    def delete(self, robot_id, key, key_signature, schema_out):
+    def delete(self, threebot_id, key, content_signature, schema_out):
```
Comment (zaibon) [inline packages/notary/actors/notary_actor.py]: We do no need a delete method after all, see comments: https://github.com/threefoldtech/home/issues/135#issuecomment-468336892

File: packages/notary/actors/notary_actor.py [hunk 4]
```diff
@@ -38,21 +54,29 @@ def get(self, key, schema_out):
...
         """
         out = schema_out.new()
-        # TODO: return json object
-        out.message = "your key is {}".format(key)
+        bcdb = self.bcdb_get()
+        for model in bcdb.get_all():
+            if model.key == key:
+                return model
+
+        out.message = "this key doesn't exist"
```
Comment (zaibon) [inline packages/notary/actors/notary_actor.py]: This should return an error not just a message. What is the proper way to return error to the client in gedis ?
Raise an exception ?

**Review by zaibon — CHANGES_REQUESTED** [review]
Added some comments. Some changes are required for the code to be inline with https://github.com/threefoldtech/home/issues/135#issuecomment-468336892

File: packages/notary/actors/notary_actor.py [hunk 5]
```diff
@@ -1,58 +1,72 @@
...
+        model = self.bcdb.models.get("threefold.grid.notary.reservation")
+        return model.get(hash=hash)
+
+
+def _hash_content(content):
+    return j.data.blake2_string(content, size=32)
+
+
+def _verify_signature(signature, verify_key):
+    verify_key.verify(signature)
```
Comment (LeeSmet) [inline packages/notary/actors/notary_actor.py]: Does verification not require both the signature and the content?

**Review by LeeSmet — CHANGES_REQUESTED** [review]
see comment on the verification

also, it seems like the key is currently the hash of the content. However this means that we need to include a timestamp in the encrypted content to ensure a unique key for things like e.g. a repeated reservation. I think it might be better to automatically set a timestamp to the time the content was uploaded in the notary, and then generate the key by hashing the content and all other data, e.g. blake2b(threebot_id|content|signature|timestamp).

File: packages/notary/actors/notary_actor.py (hunk 5 above)
Comment (zaibon) [reply packages/notary/actors/notary_actor.py]: indeed, my mistake. I'll fix. thx

**Comment by zaibon** [comment]
@LeeSmet you are write, but the timestamp is already part of what you get into the 'content' encrypted message, see https://github.com/threefoldtech/jumpscaleX/pull/216/files#diff-4843e48e342f9ae8268c7242870067f9R55

I don't think we can set the timestamp in the notary cause then we cannot verify the signature since we would change the content of the reservation.

Now this brings something else, It seems we start to have pretty much the same schema spread over different location:
- In the tfchain client for the reservation itself: https://github.com/threefoldtech/jumpscaleX/blob/cd2761e8225cdfa3c9f956a0ec6326dd7d6ab256/Jumpscale/clients/blockchain/tfchain/TFChainReservation.py#L4-L11
- In the notary to save the reservation into bcdb: https://github.com/threefoldtech/digitalmeX/blob/0e44dcd78ce78c071722e1f388f9d31f21a363fd/packages/notary/models/threefold_grid_notary_reservation.toml

These schema are somehow a bit different while still representing the same thing. Maybe we should have a central location where we store all these schema and re-use as much as possible during all the flow of reservation.

**Comment by LeeSmet** [comment]
I know the timestamp is currently part of the content, my point was if it is useful to have it there? If we don't actually do anything with the timestamp (which I don't think we do), then the only reason for the timestamp to be in the content would be for it to be part of the signature (after encryption ofcourse), which gives it a proof of integrity (i.e. it can't be tampered with by the notary). But on the other hand, it feels like we just put some useless data in the content right now to ensure a unique hash.

Regardless, I still think the key hash should at least also include the threebot_id, which we know must be authentic because otherwise the signature verification would fail.

**Comment by zaibon** [comment]
> I still think the key hash should at least also include the threebot_id

This is already the case, check the schema that is used to create the encrypted content:
https://github.com/threefoldtech/jumpscaleX/blob/cd2761e8225cdfa3c9f956a0ec6326dd7d6ab256/Jumpscale/clients/blockchain/tfchain/TFChainReservation.py#L4-L11

**Comment by LeeSmet** [comment]
yes indeed it is part of the encrypted content, but I wonder why it is part of that, since we also need it as part of the data we upload so the notary can verify the signature. So it seems we unnecessarily duplicate some data

**Comment by zaibon** [comment]
But this is what we already discussed before.
The content must contain both timestamp and threebot id to make it unique enough so user can make different reservations. 
Also I want to have these information store in the blockchain for tractability.

We indeed send the threebot id also in clear to the notary, but there is no way around that since we need it to verify the signature.

The notary is not supposed to do anything with the data, it just receive it, verify the signature and store it as is. There should be as little to no logic in the notary.

**Comment by LeeSmet** [comment]
> The content must contain both timestamp and threebot id to make it unique enough so user can make different reservations.

Indeed, reason for this being the key needs to be different. So if instead of only the content we hash both the content AND the (plaintext) 3bot id we also get a unique key without duplicating the id within the encrypted content. As for storage in the blockchain, we are going to store the key bytes, with which you will be able to get the notary data, which then includes the threebot id, content, and content signature, and you can use those 3 pieces of information to verify that everything is correct. so including the 3bot id in the encrypted content does not change anything here (after all you can't decrypt the content anyway)

**Comment by zaibon** [comment]
Ok that is indeed right that the notary will give you the threebot_id when you get the encrypted reservation. So I guess you're right and that can work like that.

**Comment by LeeSmet** [comment]
After giving this some more though over the weekend, it seems the signature is also useless for us. All benefits we get from a signature are already implied because we use Curve25519 for key derivation, which operates on both parties keypairs at once. Really the only reason for the signature is that it allows the notary to verify that the data has been created by the 3bot defined by the threebot_id field. But even then this does not necessarily guarantee that the data can be decrypted. Thoughts @zaibon ?

**Pull request merged**